connectors/protocol/connectors.py

```diff
@@ -290,6 +290,10 @@ def sync_cursor(self):
     def terminated(self):
         return self.status in (JobStatus.ERROR, JobStatus.COMPLETED, JobStatus.CANCELED)
 
+    @property
+    def checkpoint(self):
+        return self.get("checkpoint")
```

Comment:
And remind me why you don't want to to use `cursor`?

Reply:
Mostly to have them different cognitively: `cursor` is to pass state between sync jobs. `checkpoint` is to pass state within the job. I'm thinking of changing this variable to `sync_state`, how do you feel?

Reply:
I would personally prefer `checkpoint` over `sync_state`. I would worry about sync state being conflated with the health of a sync.
I think `pointer` could be a good word as well.

Reply:
I also like the original of `checkpoint`, if we're adding a new field. 

>  cursor is to pass state between sync jobs. checkpoint is to pass state within the job

I don't think I agree with this distinction, though. Checkpoint is really for in case the job fails or is interrupted, right? In which case, it _is_ for passing state between two different sync jobs (or, the same job, but two different runs of it). IMO, this feature is blurring the lines between full and incremental syncs, which I think is a good thing. I'm trying to come up with a situation where you wouldn't want this checkpoint capability to overwrite your `cursor`, and I'm struggling to come up with one. Which is why I'm thinking they should use the same field. Can you describe a counter-example?

Reply:
I tried to look at it from the following perspective:

1. Immutable "cursor" that gets never updated on the job record after it's initialised is a great audit/tracing mechanism: you can look back in time to see what point the job started from
2. The cursor gets updated only after successful full sync, meaning that if an incremental job is throttled/cancelled mid-way, the next one will try from same checkpoint. This is IMO pretty good, because the data get stale over time. If we choose to combine cursor with checkpoint, it can potentially bring bigger data inconsistencies: e.g. connector synced 50% of data from target system, hang up for a day, restarted and synced the rest 50%, whatever it means. First 50% of data will be a day older than the rest. Not sure if it's a large problem, because I don't know use-cases of our target audience.

So all in all having two fields is a bit more problematic in a short run - I'll also need to write a migration to remove the checkpoint field from mapping, so that there would be no field explosion/mapping conflicts. I don't know if all of the above make it worth to have both fields - theoretically the outcome is very similar.

I also like the blurred line between full/incremental sync - in UI they can be the same, but connector will have a single method that changes behaviour based on cursor provided.

Reply:
This gets pretty confusing as there a [connector sync_cursor](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/template-resources/src/main/resources/entsearch/connector/elastic-connectors-mappings.json#L300) and [sync job cursor](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/template-resources/src/main/resources/entsearch/connector/elastic-connectors-sync-jobs-mappings.json#L112) that references connector cursor. 
 
> I'm trying to come up with a situation where you wouldn't want this checkpoint capability to overwrite your cursor, and I'm struggling to come up with one. Which is why I'm thinking they should use the same field. Can you describe a counter-example?

Resonating with what Artem mentioned, the `sync_cursor` acts as an immutable entry point state indicating where sync jobs should start (e.g., the last sync timestamp in `google_drive`). On the other hand, checkpoints track an arbitrarily defined unit of work within the same sync operation (e.g., synced drives in gdrive). These checkpoints do not necessarily need to align with the `sync_cursor` abstraction.

This distinction becomes significant when running incremental syncs (Note: I'm aware that Artem’s changes only impact full syncs). Still, I think it's worth discussing a potential scenario with incremental syncs, as we are aware that incremental syncs are necessity for some large connector users. 

For instance, in `google_drive` incremental sync, we cannot overwrite the sync_cursor (which is a timestamp used with a delta API to fetch changes since the last sync) with a periodic checkpoint because:
- The checkpoint cannot be updated with the current timestamp since we cannot confirm that all changes up to that time have been fully synced (e.g., if the last sync was 24 hours ago and there are 10,000 new documents to process).
- In google_drive, we might need to track specific units of work such as folders, shared drives, etc., for checkpointing purposes. This is not the same as `timestamp` sync cursor
- If an incremental sync is interrupted (e.g., due to an agentless pod restart), both the last successful sync_cursor and the checkpoint are essential to correctly resume the incremental job.